gdrive

```diff
@@ -18,271 +17,412 @@ from googleapiclient.http import MediaIoBaseDownload
+        try:
+            uploader = self.google.upload(self.filepath, self.mime_type)
+            print("Uploading 0%", end='\r')
+            response = self.conclude_operation_while_logging(uploader, "Uploading")
+            if not response:
+                return
+            print("Uploaded 100% ")
+            print('File ID: %s\n' % response.get('id'))
+        except Exception as err:
+            print('An error occurred while uploading the file.')
+            print(err)
+
+
+class List(Command):
+    TYPE = "list"
+    HELP = "Browse all files or search on Google Drive."
+    FILES_PER_PAGE = 5
+
+    @staticmethod
+    def add_to_subparser(subparsers):
+        parser = subparsers.add_parser(
+            List.TYPE, 
+            help=List.HELP
+        )
+        parser.set_defaults(command=List)
+
+        parser.add_argument(
+            'file',
+            metavar='FILE',
+            nargs='?',
+            help="Name of the file to be searched (leave blank to see all files)"
+        )
+
+    def execute(self):
+        self.list_files()
+
+    def list_files(self):
+        files_found, next_page = self.search()
+        self.describe_files(files_found)
+        while next_page is not None:
+            if self.input_stop_next_page():
+                break
+            files_found, next_page = self.search(next_page)
+            self.describe_files(files_found)
+    
+    def search(self, next_page=None):
+        search_query = self.args.file
+        files_found, new_next_page = self.google.search_filename(
+            search_query,
+            self.FILES_PER_PAGE,
+            next_page
+        )
+        return files_found, new_next_page
+
+    def describe_files(self, files):
+        for file in files:
+            Utils.describe_file(file)
+
+    def input_stop_next_page(self):
+        # When Enter is pressed, the result is blank, thus false
+        return input(">>> Press Enter for the next results\n")
+
+
+class CommandParser():
+    """Starts the Commands' parsers and executes a command if the arguments are valid"""
+    COMMANDS = [Download, Upload, List]
+    NAME = "gdrive"
+    DESCRIPTION = "A script to interact with your Google Drive files"
+    
+    def __init__(self):
+        self.parser = ArgumentParser(
+            prog=self.NAME,
+            description=self.DESCRIPTION
+        )
+        subparsers = self.parser.add_subparsers()
+        for command in self.COMMANDS:
+            command.add_to_subparser(subparsers)
+        self.args = self.parser.parse_args()
+    
+    def execute_command(self):
+        if not self.is_command_valid():
+            print("Command is not valid")
+            return
+        command = self.build_command()
+        if not command.is_service_started():
+            print("Google Drive Service failed to start")
+            return
+        command.execute()
+    
+    def is_command_valid(self):
+        return self.args or self.args.command
```

Comment:
Small mistake found where an OR should be an AND
It is but a safety measure for code mistakes, so not relevant at the moment
Will fix it in a later commit

Reply:
For the record, argparse already deals with user command mistakes

> usage: gdrive [-h] {download,upload,list} ...
> gdrive: error: invalid choice: 'invalid_command' (choose from 'download', 'upload', 'list')